lib/sudo-system/src/audit.rs

```diff
@@ -19,44 +20,104 @@ fn mode(who: Category, what: Op) -> u32 {
     (what as u32) << (3 * who as u32)
 }
 
-pub fn secure_open(path: &Path) -> io::Result<File> {
-    let file = File::open(path)?;
-    let meta = file.metadata()?;
-    let permbits = meta.permissions().mode();
+pub fn secure_open(path: impl AsRef<Path>) -> io::Result<File> {
+    let mut open_options = OpenOptions::new();
+    open_options.read(true);
+    secure_open_impl(path.as_ref(), &mut open_options, false, false)
+}
+
+pub fn secure_open_cookie_file(path: impl AsRef<Path>) -> io::Result<File> {
+    let mut open_options = OpenOptions::new();
+    open_options
+        .read(true)
+        .write(true)
+        .create(true)
+        .mode(mode(Category::Owner, Op::Write) | mode(Category::Owner, Op::Read));
+    secure_open_impl(path.as_ref(), &mut open_options, true, true)
+}
+
+fn checks(path: &Path, meta: Metadata) -> io::Result<()> {
     let error = |msg| Error::new(ErrorKind::PermissionDenied, msg);
 
+    let path_mode = meta.permissions().mode();
     if meta.uid() != 0 {
         Err(error(format!("{} must be owned by root", path.display())))
-    } else if meta.gid() != 0 && (permbits & mode(Category::Group, Op::Write) != 0) {
+    } else if meta.gid() != 0 && (path_mode & mode(Category::Group, Op::Write) != 0) {
         Err(error(format!(
             "{} cannot be group-writable",
             path.display()
         )))
-    } else if permbits & mode(Category::World, Op::Write) != 0 {
+    } else if path_mode & mode(Category::World, Op::Write) != 0 {
         Err(error(format!(
             "{} cannot be world-writable",
             path.display()
         )))
     } else {
-        Ok(file)
+        Ok(())
+    }
+}
+
+fn secure_open_impl(
+    path: &Path,
+    open_options: &mut OpenOptions,
+    check_parent_dir: bool,
+    create_parent_dirs: bool,
+) -> io::Result<File> {
+    let error = |msg| Error::new(ErrorKind::PermissionDenied, msg);
+    if check_parent_dir || create_parent_dirs {
+        if let Some(parent_dir) = path.parent() {
+            // if we should create parent dirs and it does not yet exist, create it
+            if create_parent_dirs && !parent_dir.exists() {
+                DirBuilder::new()
+                    .recursive(true)
+                    .mode(
+                        mode(Category::Owner, Op::Write)
```

Comment:
I think I can see a beautification step here (having foisted this data type on us), but I'll do that in a follow-up